e2e/test_reconnection.sh

```diff
@@ -22,67 +22,116 @@ case "$1" in
-kill $sl_pid
+check_and_kill "$sl_pids"
 echo "Killing Superlink"
 sleep 3
 
 # Restart superlink, the clients should now be able to reconnect to it
-timeout 2m flower-superlink --insecure $db_arg $rest_arg --isolation="process" &
-sl_pid=$!
+timeout 10m flower-superlink --insecure $db_arg $rest_arg --executor flwr.superexec.deployment:executor \
+  2>&1 | tee flwr_output.log &
+sl_pids=$(pgrep -f "flower-superlink")
 echo "Restarting Superlink"
 sleep 20
 
-# Kill first client, this should send a DeleteNode message to the Superlink
+# Kill second client, this should send a DeleteNode message to the Superlink
 kill $cl1_pid
-echo "Killing first client"
-sleep 3
+echo "Killing second client"
+sleep 5
 
-# Starting new client, this is so we have enough clients to start the server-app
-timeout 2m flower-supernode ./ --insecure $rest_arg --superlink $server_address &
+# Starting new client, this is so we have enough clients to execute `flwr run`
+timeout 10m flower-supernode ./ --insecure $rest_arg --superlink $server_address \
+  --isolation="subprocess" --clientappio-api-address "localhost:9094" &
 cl1_pid=$!
 echo "Starting new client"
 sleep 5
 
-# We start the server-app to begining the training
-timeout 2m flower-server-app ./ $rest_arg --superlink $server_app_address &
-pid=$!
-echo "Starting server-app to start training"
+# We execute `flwr run` to begin the training
+timeout 2m flwr run "." e2e &
+echo "Executing flwr run to start training"
+sleep 10
 
-# Kill first client as soon as the training starts,
-# the server-app should just receive a failure in this case and continue the rounds
-# when enough clients are connected
+# Kill first client as soon as the training starts, the flwr-serverapp should just 
+# receive a failure in this case and continue the rounds when enough clients are 
+# connected
 kill $cl1_pid
 echo "Killing first client"
-sleep 1
+sleep 3
 
 # Restart first client so enough clients are connected to continue the FL rounds
-timeout 2m flower-supernode ./ --insecure $rest_arg --superlink $server_address &
+timeout 5m flower-supernode ./ --insecure $rest_arg --superlink $server_address \
+  --isolation="subprocess" --clientappio-api-address "localhost:9094" &
 cl1_pid=$!
 echo "Starting new client"
+sleep 5
 
-wait $pid
-res=$?
+# Initialize a flag to track if training is successful
+found_success=false
+timeout=120  # Timeout after 120 seconds
+elapsed=0
 
-if [[ "$res" = "0" ]];
-  then echo "Training worked correctly"; kill $cl1_pid; kill $cl2_pid; kill $sl_pid;
-  else echo "Training had an issue" && exit 1;
-fi
+# Check for "Success" in a loop with a timeout
+while [ "$found_success" = false ] && [ $elapsed -lt $timeout ]; do
+    if grep -q "Run finished" flwr_output.log; then
+        echo "Training worked correctly!"
+        found_success=true
+        kill $cl1_pid; kill $cl2_pid
+        sleep 3
+        check_and_kill "$cl1_pids"
+        check_and_kill "$sl_pids"
```

Comment:
Just a q question, where do we set `cl1_pids` and `sl_pids`?

Reply:
You're right, `cl1_pids` is not set. I'll remove it.